File: Leetcode/Python/1014. Best Sightseeing Pair/solution.py
# https://leetcode.com/problems/best-sightseeing-pair/description

class Solution:
    def maxScoreSightseeingPair(self, values: List[int]) -> int:
        max_left = values[0]
        max_value = 0

        # Maximize values[i] + i
        for idx in range(1, len(values)):
            max_value = max(max_value, max_left + values[idx] - idx)
            max_left = max(max_left, values[idx] + idx)

        return max_value

File: Leetcode/Typescript/1014. Best Sightseeing Pair/solution.ts
// https://leetcode.com/problems/best-sightseeing-pair/description

function maxScoreSightseeingPair(values: number[]): number {
  let maxLeft = values[0];
  let maxValue = 0;
  // Maximize values[i] + i
  for (let idx = 1; idx < values.length; idx++) {
    maxValue = Math.max(maxValue, maxLeft + values[idx] - idx);
    maxLeft = Math.max(maxLeft, values[idx] + idx);
  }

  return maxValue;
}
